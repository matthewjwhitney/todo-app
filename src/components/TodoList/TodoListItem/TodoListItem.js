import {
  Box,
  lighten,
  ListItem,
  ListItemIcon,
  ListItemText,
  makeStyles,
  Typography,
} from "@material-ui/core";
import { getMonthDayYearFromUTCDate, getIsOverdue } from "../../../utils";
import BlackCheckbox from "../../BlackCheckbox";

const useStyles = makeStyles((theme) => ({
  listItemIcon: {
    minWidth: theme.spacing(1),
  },
  listItemGrey: {
    backgroundColor: theme.palette.grey[300],
    margin: theme.spacing(1),
    "&:hover": {
      backgroundColor: theme.palette.grey[200],
    },
    "&:focus": {
      backgroundColor: theme.palette.grey[200],
    },
  },
  listItemGreen: {
    backgroundColor: lighten(theme.palette.success.light, 0.4),
    margin: theme.spacing(1),
    "&:hover": {
      backgroundColor: lighten(theme.palette.success.light, 0.6),
    },
    "&:focus": {
      backgroundColor: lighten(theme.palette.success.light, 0.6),
    },
  },
  listItemRed: {
    backgroundColor: lighten(theme.palette.error.light, 0.4),
    margin: theme.spacing(1),
    "&:hover": {
      backgroundColor: lighten(theme.palette.error.light, 0.6),
    },
    "&:focus": {
      backgroundColor: lighten(theme.palette.error.light, 0.6),
    },
  },
  ListItemTextCompleted: {
    textDecoration: "line-through",
  },
}));

const TodoListItem = ({
  id,
  isComplete,
  description,
  dueDate,
  onClickCheckbox,
}) => {
  const classes = useStyles();
  const isOverdue = getIsOverdue(dueDate);

  const getListItemStyle = () => {
    if (isComplete) {
      return classes.listItemGreen;
    } else if (isOverdue) {
      return classes.listItemRed;
    } else {
      return classes.listItemGrey;
    }
  };

  return (
    <ListItem
      classes={{ root: getListItemStyle() }}
      dense
      button
      data-testid="todo-list-item-button"
      onClick={() => onClickCheckbox(id)}
    >
      <ListItemIcon className={classes.listItemIcon}>
        <BlackCheckbox
          edge="start"
          checked={isComplete}
          disableRipple
          tabIndex={-1}
          inputProps={{ "aria-labelledby": `checkbox-list-label-${id}` }}
        />
      </ListItemIcon>
      <ListItemText
        id={`checkbox-list-label-${id}`}
        className={isComplete ? classes.ListItemTextCompleted : ""}
        primary={
          <Typography variant="h5" component="span">
            {description}
          </Typography>
        }
      />
      {dueDate && (
        <Box edge="end" border={1} p={0.5} m={0.5}>
          <Typography variant="h6" component="span">
            {getMonthDayYearFromUTCDate(dueDate)}
          </Typography>
        </Box>
      )}
    </ListItem>
  );
};

export default TodoListItem;
