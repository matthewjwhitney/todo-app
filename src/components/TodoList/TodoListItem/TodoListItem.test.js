import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import TodoListItem from "./TodoListItem";

const overdueProps = {
  description: "Feed the cat",
  dueDate: "2020-06-24T15:45:00.000Z",
  id: "5",
  isComplete: false,
  onClickCheckbox: jest.fn(),
};

const regularProps = {
  description: "Walk the dog",
  dueDate: null,
  id: "4",
  isComplete: false,
  onClickCheckbox: jest.fn(),
};

const completeProps = {
  description: "File 2020 Taxes",
  dueDate: "2020-03-10T17:50:44.673Z",
  id: "1",
  isComplete: true,
  onClickCheckbox: jest.fn(),
};

it("should render an overdue todo correctly", () => {
  const renderedComponent = render(<TodoListItem {...overdueProps} />);
  expect(renderedComponent).toMatchSnapshot();
});

it("should render a regular todo correctly", () => {
  const renderedComponent = render(<TodoListItem {...regularProps} />);
  expect(renderedComponent).toMatchSnapshot();
});

it("should render a completed todo correctly", () => {
  const renderedComponent = render(<TodoListItem {...completeProps} />);
  expect(renderedComponent).toMatchSnapshot();
});

it("should perform the onClickCheckbox function if the button is clicked", () => {
  render(<TodoListItem {...regularProps} />);
  const button = screen.getByTestId("todo-list-item-button");
  fireEvent.click(button);
  expect(regularProps.onClickCheckbox).toHaveBeenCalledTimes(1);
});
