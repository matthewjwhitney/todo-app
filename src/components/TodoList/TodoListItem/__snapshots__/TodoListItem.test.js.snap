// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a completed todo correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root MuiListItem-root makeStyles-listItemGreen-26 MuiListItem-dense MuiListItem-gutters MuiListItem-button"
        data-testid="todo-list-item-button"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root makeStyles-listItemIcon-24"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-31 MuiCheckbox-root Component-root-29 PrivateSwitchBase-checked-32 Mui-checked Component-checked-30 MuiIconButton-edgeStart"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                aria-labelledby="checkbox-list-label-1"
                checked=""
                class="PrivateSwitchBase-input-34"
                data-indeterminate="false"
                tabindex="-1"
                type="checkbox"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                />
              </svg>
            </span>
          </span>
        </div>
        <div
          class="MuiListItemText-root makeStyles-ListItemTextCompleted-28 MuiListItemText-dense"
          id="checkbox-list-label-1"
        >
          <span
            class="MuiTypography-root MuiTypography-h5"
          >
            File 2020 Taxes
          </span>
        </div>
        <div
          class="MuiBox-root MuiBox-root-35"
          edge="end"
        >
          <span
            class="MuiTypography-root MuiTypography-h6"
          >
            3/10/2020
          </span>
        </div>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      aria-disabled="false"
      class="MuiButtonBase-root MuiListItem-root makeStyles-listItemGreen-26 MuiListItem-dense MuiListItem-gutters MuiListItem-button"
      data-testid="todo-list-item-button"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiListItemIcon-root makeStyles-listItemIcon-24"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-31 MuiCheckbox-root Component-root-29 PrivateSwitchBase-checked-32 Mui-checked Component-checked-30 MuiIconButton-edgeStart"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              aria-labelledby="checkbox-list-label-1"
              checked=""
              class="PrivateSwitchBase-input-34"
              data-indeterminate="false"
              tabindex="-1"
              type="checkbox"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
          </span>
        </span>
      </div>
      <div
        class="MuiListItemText-root makeStyles-ListItemTextCompleted-28 MuiListItemText-dense"
        id="checkbox-list-label-1"
      >
        <span
          class="MuiTypography-root MuiTypography-h5"
        >
          File 2020 Taxes
        </span>
      </div>
      <div
        class="MuiBox-root MuiBox-root-35"
        edge="end"
      >
        <span
          class="MuiTypography-root MuiTypography-h6"
        >
          3/10/2020
        </span>
      </div>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`should render a regular todo correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root MuiListItem-root makeStyles-listItemGrey-14 MuiListItem-dense MuiListItem-gutters MuiListItem-button"
        data-testid="todo-list-item-button"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root makeStyles-listItemIcon-13"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-20 MuiCheckbox-root Component-root-18 MuiIconButton-edgeStart"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                aria-labelledby="checkbox-list-label-4"
                class="PrivateSwitchBase-input-23"
                data-indeterminate="false"
                tabindex="-1"
                type="checkbox"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
          </span>
        </div>
        <div
          class="MuiListItemText-root MuiListItemText-dense"
          id="checkbox-list-label-4"
        >
          <span
            class="MuiTypography-root MuiTypography-h5"
          >
            Walk the dog
          </span>
        </div>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      aria-disabled="false"
      class="MuiButtonBase-root MuiListItem-root makeStyles-listItemGrey-14 MuiListItem-dense MuiListItem-gutters MuiListItem-button"
      data-testid="todo-list-item-button"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiListItemIcon-root makeStyles-listItemIcon-13"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-20 MuiCheckbox-root Component-root-18 MuiIconButton-edgeStart"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              aria-labelledby="checkbox-list-label-4"
              class="PrivateSwitchBase-input-23"
              data-indeterminate="false"
              tabindex="-1"
              type="checkbox"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
        </span>
      </div>
      <div
        class="MuiListItemText-root MuiListItemText-dense"
        id="checkbox-list-label-4"
      >
        <span
          class="MuiTypography-root MuiTypography-h5"
        >
          Walk the dog
        </span>
      </div>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`should render an overdue todo correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root MuiListItem-root makeStyles-listItemRed-4 MuiListItem-dense MuiListItem-gutters MuiListItem-button"
        data-testid="todo-list-item-button"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root makeStyles-listItemIcon-1"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-8 MuiCheckbox-root Component-root-6 MuiIconButton-edgeStart"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                aria-labelledby="checkbox-list-label-5"
                class="PrivateSwitchBase-input-11"
                data-indeterminate="false"
                tabindex="-1"
                type="checkbox"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
          </span>
        </div>
        <div
          class="MuiListItemText-root MuiListItemText-dense"
          id="checkbox-list-label-5"
        >
          <span
            class="MuiTypography-root MuiTypography-h5"
          >
            Feed the cat
          </span>
        </div>
        <div
          class="MuiBox-root MuiBox-root-12"
          edge="end"
        >
          <span
            class="MuiTypography-root MuiTypography-h6"
          >
            6/24/2020
          </span>
        </div>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      aria-disabled="false"
      class="MuiButtonBase-root MuiListItem-root makeStyles-listItemRed-4 MuiListItem-dense MuiListItem-gutters MuiListItem-button"
      data-testid="todo-list-item-button"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiListItemIcon-root makeStyles-listItemIcon-1"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-8 MuiCheckbox-root Component-root-6 MuiIconButton-edgeStart"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              aria-labelledby="checkbox-list-label-5"
              class="PrivateSwitchBase-input-11"
              data-indeterminate="false"
              tabindex="-1"
              type="checkbox"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
        </span>
      </div>
      <div
        class="MuiListItemText-root MuiListItemText-dense"
        id="checkbox-list-label-5"
      >
        <span
          class="MuiTypography-root MuiTypography-h5"
        >
          Feed the cat
        </span>
      </div>
      <div
        class="MuiBox-root MuiBox-root-12"
        edge="end"
      >
        <span
          class="MuiTypography-root MuiTypography-h6"
        >
          6/24/2020
        </span>
      </div>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
